lib/web/desktop_playback.go

```diff
@@ -17,36 +17,68 @@ limitations under the License.
 package web
 
 import (
+	"context"
 	"net/http"
 
 	"github.com/gravitational/trace"
 	"github.com/julienschmidt/httprouter"
 	"golang.org/x/net/websocket"
 
+	"github.com/gravitational/teleport/lib/player"
 	"github.com/gravitational/teleport/lib/reversetunnelclient"
+	"github.com/gravitational/teleport/lib/session"
 	"github.com/gravitational/teleport/lib/web/desktop"
 )
 
 func (h *Handler) desktopPlaybackHandle(
 	w http.ResponseWriter,
 	r *http.Request,
 	p httprouter.Params,
-	ctx *SessionContext,
+	sctx *SessionContext,
 	site reversetunnelclient.RemoteSite,
 ) (interface{}, error) {
 	sID := p.ByName("sid")
 	if sID == "" {
-		return nil, trace.BadParameter("missing sid in request URL")
+		return nil, trace.BadParameter("missing session ID in request URL")
 	}
 
-	clt, err := ctx.GetUserClient(r.Context(), site)
+	clt, err := sctx.GetUserClient(r.Context(), site)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	websocket.Handler(func(ws *websocket.Conn) {
-		defer h.log.Debug("desktopPlaybackHandle websocket handler returned")
-		desktop.NewPlayer(sID, ws, clt, h.log).Play(r.Context())
+		ws.PayloadType = websocket.BinaryFrame
+
+		player, err := player.New(&player.Config{
+			Clock:     h.clock,
+			Log:       h.log,
+			SessionID: session.ID(sID),
+			Streamer:  clt,
+		})
+		if err != nil {
+			h.log.Errorf("couldn't create player for session %v: %v", sID, err)
+			ws.Write([]byte(`{"message": "error", "errorText": "Internal server error"}`))
+			return
+		}
+
+		defer player.Close()
+
+		ctx, cancel := context.WithCancel(r.Context())
+		defer cancel()
+
+		go func() {
+			defer cancel()
+			desktop.ReceivePlaybackActions(h.log, ws, player)
+		}()
+
+		go func() {
+			defer cancel()
```

Comment:
This ensures that the other loop (in `ReceivePlaybackActions`) always breaks if this goroutine returns.

```suggestion
			defer cancel()
			defer ws.Close()
```